lib/geometry/polygon.nit

```diff
@@ -0,0 +1,428 @@
+	# poly.anti_clock_sort
+	# assert poly.is_convex
+	# ~~~
+	fun is_convex: Bool do
+		var prev = points[points.length - 2]
+		var curr = points[points.length - 1]
+		var next = points[0]
+		var is_ccw = turn_left(prev, curr, next)
+		for i in [1..points.length[ do
+			prev = curr
+			curr= next
+			next = points[i]
+			if turn_left(prev ,curr, next) != is_ccw then return false
+		end
+		return true
+	end
+
+	# Check if `p` is in the polygon
+	# ~~~
+	# var p1 = new Point[Float](0.0, 0.0)
+	# var p2 = new Point[Float](5.0, 0.0)
+	# var p3 = new Point[Float](0.0, 5.0)
+	# var p4 = new Point[Float](5.0, 5.0)
+	# var p5 = new Point[Float](2.5, 2.5)
+	# var arr = new Array[Point[Float]].with_items(p1, p2, p3, p4)
+	# var poly = new ConvexPolygon.with_vertices(arr)
+	# poly.anti_clock_sort
+	# assert poly.contain(p5)
+	# ~~~
+	fun contain(p : Point[Float]): Bool do
+		var prev = points[points.length - 1]
+		var curr = p
+		var next = points[0]
+		var is_ccw = turn_left(prev, curr, next)
+		for i in [1..points.length[ do
+			prev = next
+			next = points[i]
+			if turn_left(prev, curr, next) != is_ccw then return false
+		end
+		print points.length
+		return true
+	end
+
+	# Check if the order of the points in the polygon is counter-clockwise
+	# The vertices in the polygon need to be sorted
+	# ~~~
+	# var p1 = new Point[Float](0.0, 0.0)
+	# var p2 = new Point[Float](5.0, 0.0)
+	# var p3 = new Point[Float](0.0, 5.0)
+	# var p4 = new Point[Float](5.0, 5.0)
+	# var arr = new Array[Point[Float]].with_items(p1, p2, p3, p4)
+	# var poly = new ConvexPolygon.with_vertices(arr)
+	# poly.anti_clock_sort
+	# assert poly.is_ccw
+	# ~~~ 
+	fun is_ccw: Bool do
+		var min = points[0].y
+		var min_index = 0
+		for i in [1..points.length - 1[ do
+			if points[i].y < min then
+				min = points[i].y
+				min_index = i
+			end
+		end
+		var prev = points[(min_index - 1 + points.length) % points.length]
+		var next = points[(min_index + 1) % points.length]
+		return not turn_left(prev, points[min_index], next)
+	end
+
+	# Remove  `p` from the vertices, keeping at least 3 vertices
+	fun delete_vertex(p: Point[Float]) do
+		if points.length > 3 then
+			points.remove(p)
+		end
+	end
+
+	# Add a vertice to the polygon
+	# ~~~
+	# var p1 = new Point[Float](0.0, 0.0)
+	# var p2 = new Point[Float](5.0, 0.0)
+	# var p3 = new Point[Float](0.0, 5.0)
+	# var p4 = new Point[Float](5.0, 5.0)
+	# var arr = new Array[Point[Float]].with_items(p1, p2, p3, p4)
+	# var poly = new ConvexPolygon.with_vertices(arr)
+	# var p5 = new Point[Float](2.5, 7.5)
+	# assert poly.add_vertex(p5)
+	# ~~~ 
+	fun add_vertex(p: Point[Float]): Bool do
+		assert points.length >= 3
+		var temp_list = points.clone
+		temp_list.add(p)
+		var temp_polygon = new ConvexPolygon.with_vertices(temp_list)
+		temp_polygon.anti_clock_sort
+		if temp_polygon.is_convex then
+			points = temp_polygon.points
+			return true
+		else
+			return false
+		end
+	end
+end
+
+private class Projection
```

Comment:
I'm not sure I understand the use of this class, though private, a small synopsis to explain its use might be nice for future maintainers